pacman/model/routing_info/routing_info.py

```diff
@@ -41,46 +43,175 @@ def add_routing_info(self, info: VertexRoutingInfo):
         :rtype: VertexRoutingInfo
+        :raise KeyError:
+            If the vertex/partition_id combination is not in the routing
+            information
         """
-        return self._info.get((vertex, partition_id))
+        return self._info[vertex][partition_id]
 
+    @deprecated(reason="This method is unsafe, since it doesn't determine "
+                       "whether the info is missing because there is no "
+                       "outgoing edge, or if the outgoing edge is in another "
+                       "partition and the name is wrong. "
+                       "Use a combination of "
+                       "get_key_from, "
+                       "get_partitions_from, "
+                       "has_info_from, "
+                       "or get_single_key_from")
     def get_first_key_from_pre_vertex(
             self, vertex: AbstractVertex, partition_id: str) -> Optional[int]:
         """
         Get the first key for the partition starting at a vertex.
 
+        :param AbstractVertex vertex: The vertex which the partition starts at
+        :param str partition_id:
+            The ID of the partition for which to get the routing information
+        :return: The routing key of the partition
+        :rtype: int or None
+        """
+        if vertex not in self._info:
+            return None
+        info = self._info[vertex]
+        if partition_id not in info:
+            return None
+        return info[partition_id].key
+
+    def get_key_from(
+            self, vertex: AbstractVertex, partition_id: str) -> int:
+        """
+        Get the first key for the partition starting at a vertex.
+
         :param AbstractVertex vertex: The vertex which the partition starts at
         :param str partition_id:
             The ID of the partition for which to get the routing information
         :return: The routing key of the partition
         :rtype: int
+        :raise KeyError:
+            If the vertex/partition_id combination is not in the routing
+            information
+        """
+        return self._info[vertex][partition_id].key
+
+    def get_partitions_from(
+            self, vertex: AbstractVertex) -> Iterable[str]:
+        """
+        Get the outgoing partitions from a vertex.
+
+        :param AbstractVertex vertex: The vertex to search for
+        """
+        return self._info[vertex].keys()
+
+    def has_info_from(
+            self, vertex: AbstractVertex, partition_id: str) -> bool:
+        """
+        Check if there is routing information for a given vertex.
+
+        :param AbstractVertex vertex: The vertex to search for
+        :param str partition_id:
+            The ID of the partition for which to get the routing information
+        :rtype: bool
+        """
+        if vertex not in self._info:
+            return False
+        info = self._info[vertex]
+        return partition_id in info
+
+    def check_info_from(
```

Comment:
While this method is fine I would rather see this functionality moved to add_routing_info

If the  info.vertex knows the allowed_partition_ids before any edges or Projections are added. Check that in add_routing_info which is called when a Projection/ edge is added so offers fail fast!

Reply:
From a user perspective, this will likely not make much of a difference, since I would assume this happens only during a run call.  It will be earlier in the process, but not at time of edge creation, so not much better.  Adding a check at the time of edge creation makes more sense if this is going to be done, but is also perhaps for another PR!

Reply:
(That is, add_routing_info still doesn't happen during Projection / edge creation, only in run)

Reply:
Note also that in reality a *check* should be done at the add time, since a sequence of adding projections may change the outcome in that case e.g. for one vertex, the first partition id "checked" can be the one it will use for its outgoing key, and then any additional ones can be either failures or accepted depending on if the vertex can send with more than one key or not.  It might then accept the names of only certain partitions if it needs to know which partitions are for which reason as well.  A receiver vertex could then also "check" that the given partition id is OK, which could then check if it has already got a key from that vertex, or look into which projection is using the key or ... lots of other options.  This could be quite tricky to implement in general, so probably an implementation per-vertex might be needed, and so definitely doing this in another PR would be recommended!

The idea of the use of the calls being changed is that it is not as complicated for existing vertices.  Yes, it means you don't know until later if you have gone wrong, but given that this is unlikely to be much of a problem most of the time, I don't think that is a big issue!

Reply:
was just an idea not critical.   Can be looked at again when the multi partition stuff comes in